filterExprParser.go

```diff
@@ -0,0 +1,691 @@
+// Copyright 2018-2019 Couchbase, Inc. All rights reserved.
+
+package gojsonsm
+
+import (
+	"fmt"
+	"github.com/alecthomas/participle"
+	"strings"
+)
+
+// EBNF Grammar describing the parser
+
+// FilterExpression 		= AndCondition { "OR" AndCondition }
+// AndCondition 			= Condition { "AND" Condition }
+// Condition				= [ "NOT" ] Operand
+// Operand					= { Paren } BooleanExpr | ( LHS ( CheckOp | ( CompareOp RHS) ) ) { Paren }
+// Paren			    	= OpenParen | CloseParen
+// BooleanExpr				= Boolean | BooleanFuncExpr
+// LHS 						= ConstFuncExpr | Field | Value
+// RHS 						= ConstFuncExpr | Value | Field
+// CompareOp				= "=" | "==" | "!=" | "<>" | ">" | ">=" | "<" | "<="
+// CheckOp					= "EXISTS" | ( "IS" [ "NOT" ] ( NULL | MISSING ) )
+// Field					= OnePath { "." OnePath }
+// OnePath					= ( PathFuncExpression | @String | @Ident ){ ArrayIndex }
+// ArrayIndex				= "[" @"-" @Int "]"
+// Value					= "@String"
+// ConstFuncExpr			= ConstFuncNoArg | ConstFuncOneArg | ConstFuncTwoArgs
+// ConstFuncNoArg			= ( ConstFuncNoArgName "(" ")" )
+// ConstFuncNoArgName 		= "PI" | "E"
+// ConstFuncOneArg 			= ( ConstFuncOneArgName "(" ConstFuncArgument ")" )
+// ConstFuncOneArgName  	= "ABS" | "ACOS"...
+// ConstFuncTwoArgs			= ( ConstFuncTwoArgsName "(" ConstFuncArgument "," ConstFuncArgument ")" )
+// ConstFuncTwoArgsName 	 = "ATAN2" | "POW"
+// ConstFuncArgument		= Field | Value | ConstFuncExpr
+// PathFuncExpression		= OnePathFuncNoArg
+// OnePathFuncNoArg			= ( OnePathFuncNoArgName "(" ")" )
+// OnePathFuncNoArgName		= "META"
+// BooleanFuncExpr			= BooleanFuncTwoArgs
+// BooleanFuncTwoArgs		= ( BooleanFuncTwoArgsName "(" ConstFuncArgument "," ConstFuncArgument ")" )
+// BooleanFuncTwoArgsName 	= "REGEX_CONTAINS"
```

Comment:
Good catch thanks.

Reply:
AFA the string literal goes, for the sake of capturing I've done something called FEConstFuncArgumentRHS, which will take away the field capture. I still want to have a subFunc expression just in case for future compatibility though. For REGEXP_CONTAINS, I will do subsequent checks.